src/dc_msg.rs

```diff
@@ -831,98 +832,70 @@ pub unsafe fn dc_msg_get_summarytext(
+    let ret: String;
+    let mut prefix = "".to_string();
+    let mut append_text = true;
     match type_0 {
-        Viewtype::Image => prefix = context.stock_str(StockMessage::Image).strdup(),
-        Viewtype::Gif => prefix = context.stock_str(StockMessage::Gif).strdup(),
-        Viewtype::Video => prefix = context.stock_str(StockMessage::Video).strdup(),
-        Viewtype::Voice => prefix = context.stock_str(StockMessage::VoiceMessage).strdup(),
+        Viewtype::Image => prefix = context.stock_str(StockMessage::Image).to_string(),
+        Viewtype::Gif => prefix = context.stock_str(StockMessage::Gif).to_string(),
+        Viewtype::Video => prefix = context.stock_str(StockMessage::Video).to_string(),
+        Viewtype::Voice => prefix = context.stock_str(StockMessage::VoiceMessage).to_string(),
         Viewtype::Audio | Viewtype::File => {
             if param.get_int(Param::Cmd) == Some(6) {
-                prefix = context.stock_str(StockMessage::AcSetupMsgSubject).strdup();
-                append_text = 0i32
+                prefix = context.stock_str(StockMessage::AcSetupMsgSubject).to_string();
+                append_text = false
             } else {
-                pathNfilename = param
-                    .get(Param::File)
-                    .unwrap_or_else(|| "ErrFilename")
-                    .strdup();
-                value = dc_get_filename(pathNfilename);
-                let label = CString::new(
-                    context
+                let value;
+                unsafe {
+                    let pathNfilename = param
+                        .get(Param::File)
+                        .unwrap_or_else(|| "ErrFilename").strdup();
+                    value = as_str(dc_get_filename(pathNfilename));
+                    free(pathNfilename as *mut libc::c_void);
+                }
+                let label = context
                         .stock_str(if type_0 == Viewtype::Audio {
                             StockMessage::Audio
                         } else {
                             StockMessage::File
-                        })
-                        .as_ref(),
-                )
-                .unwrap();
-                prefix = dc_mprintf(
-                    b"%s \xe2\x80\x93 %s\x00" as *const u8 as *const libc::c_char,
-                    label.as_ptr(),
-                    value,
-                )
+                        });
+                prefix = format!("{} – {}", label, value)
             }
         }
         _ => {
             if param.get_int(Param::Cmd) == Some(9) {
-                prefix = context.stock_str(StockMessage::Location).strdup();
-                append_text = 0;
+                prefix = context.stock_str(StockMessage::Location).to_string();
+                append_text = false;
             }
         }
     }
-    if 0 != append_text
-        && !prefix.is_null()
-        && !text.is_null()
-        && 0 != *text.offset(0isize) as libc::c_int
```

Comment:
Not really sure if i can just skip those weird offset checks. Any ideas?

Reply:
you should keep a check for `!text.is_empty()` (which is the the `offset` call)